### PR TITLE
Clarify ExternalTrafficPolicy/InternalTrafficPolicy definitions

### LINKED ISSUE #108526: definition of "internal" vs "external" service traffic, behavior of LB IP short-circuit
Service traffic can be "internal" or "external" but we never explicitly define exactly what these mean.

[The docs](https://kubernetes.io/docs/concepts/services-networking/service/#external-traffic-policy) say:

> You can set the `spec.externalTrafficPolicy` field to control how traffic _**from external sources**_ is routed.

(emphasis added)

The [internalTrafficPolicy KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2086-service-internal-traffic-policy) says:

> This field will be independent from `externalTrafficPolicy`. In other words, `internalTrafficPolicy` only applies to traffic originating _**from internal sources**_

(emphasis added)

These both describe a "source" model: `internalTrafficPolicy` applies to traffic that comes from an internal source (pod, node) and `externalTrafficPolicy` applies to traffic that comes from an external source (outside cluster).

In discussion in the SIG on 2022-03-03, it seemed that most people assumed / believed / desired that we used a "destination" model: `internalTrafficPolicy` applies to traffic that is sent to an internal destination (ClusterIP), while `externalTrafficPolicy` applies to traffic that is sent to an external destination (NodePorts, ExternalIPs, and LoadBalancer IPs).

FWIW, although the docs imply the "source" model, it's not clear that this really means that we actually _intended_ that; if you're only thinking about the "normal" cases (pod to cluster IP, cluster-external to NodePort/ExternalIP/LoadBalancer) then the two models are the same. Eg, the API docs for `internalTrafficPolicy` say:

```
	// InternalTrafficPolicy specifies if the cluster internal traffic
	// should be routed to all endpoints or node-local endpoints only.
```

where "cluster internal" implies to me that they were only really considering the case of internal-to-internal traffic.

----

The current kube-proxy iptables implementation is a mash-up of the two models which also depends on whether you have a functioning LocalTrafficDetector. (ie, if you passed either `--detect-local-mode ...` to explicitly tell it how to detect local traffic, or if you passed `--cluster-cidr ...` which implicitly lets it detect that):

(Ⓢ/~Ⓢ~ = right/wrong according to the "source" model, Ⓓ/~Ⓓ~ = right/wrong according to the "destination" model)

|source|destination|w/ local detector|no local detector|
|----|----|----|----|
|pod|ClusterIP|internal Ⓢ Ⓓ|internal Ⓢ Ⓓ|
|pod|same-node NodePort|internal Ⓢ ~Ⓓ~|external ~Ⓢ~ Ⓓ|
|pod|remote NodePort|external ~Ⓢ~ Ⓓ|external ~Ⓢ~ Ⓓ|
|pod|ExternalIP|internal Ⓢ ~Ⓓ~|external ~Ⓢ~ Ⓓ but sometimes broken|
|pod|LoadBalancer IP|internal Ⓢ ~Ⓓ~|external ~Ⓢ~ Ⓓ but sometimes broken|
|node|ClusterIP|internal Ⓢ Ⓓ|internal Ⓢ Ⓓ|
|node|same-node NodePort|internal Ⓢ ~Ⓓ~|external ~Ⓢ~ Ⓓ|
|node|remote NodePort|external ~Ⓢ~ Ⓓ|external ~Ⓢ~ Ⓓ|
|node|ExternalIP|internal Ⓢ ~Ⓓ~|external ~Ⓢ~ Ⓓ|
|node|LoadBalancer IP|internal Ⓢ ~Ⓓ~|external ~Ⓢ~ Ⓓ|
|external|ClusterIP|internal ~Ⓢ~ Ⓓ|internal ~Ⓢ~ Ⓓ but sometimes broken|
|external|NodePort|external Ⓢ Ⓓ|external Ⓢ Ⓓ|
|external|ExternalIP|external Ⓢ Ⓓ|external Ⓢ Ⓓ|
|external|LoadBalancer IP|external Ⓢ Ⓓ|external Ⓢ Ⓓ|

The reason some of the cases are inconsistent between "w/ local detector" and "no local detector" is that "pod/node to NodePort/ExternalIP/LBIP" connections get short-circuited to the "pod/node to ClusterIP" case, so that if the service is `externalTrafficPolicy: Local`, the behavior doesn't depend on whether there are local endpoints or not. This can always happen for node-to-X connection, but for pod-to-X we can only do it when we have a LocalTrafficDetector.

However, the short-circuiting means that if you have a service which has Local _internal_ traffic policy, then pod-to-LB-IP connections to it won't work from nodes with no endpoints. I am pretty sure that was not the intent; the point of short-circuiting is not that the traffic should be treated as "internal" rather than "external", it's that pod-to-LB-IP traffic should behave the same way as as external-to-LB-IP traffic does, meaning the client isn't responsible for thinking about which nodes have local endpoints, because the load balancer does that for them.

----

Given

1. The "working LocalTrafficDetector" case does not consistently obey either the "source" model or the "destination" model and its behavior is fairly inexplicable.
2. The "no LocalTrafficDetector" case consistently obeys the "destination" model
3. Almost everyone in the SIG seemed to prefer the "destination" model to the "source" model

I think we should declare that the "destination" model is correct, clarify the docs and the Internal Traffic Policy KEP, and fix iptables kube-proxy.

To avoid breaking people using pod/node-to-LB-IP, this also requires fixing the short circuit so that instead of short-circuiting "external" traffic to the "internal" codepath, we instead treat it internally as a third kind of traffic ("traffic that we have to pretend left the cluster and then came back in in a manner that is consistent with `externalTrafficPolicy`").

----

This would mean:
- The "with local detector" and "no local detector" cases would agree in all cases about "internal" vs "external", though the "no local detector" case would still have problems implementing some of the cases correctly.
- Pod/Node to same-node NodePort would be considered "external" rather than "internal". IIRC Tim suggested that this isn't a big deal, because it's unlikely anyone making pod-to-nodeport connections is also using `externalTrafficPolicy: Local`
    - (EDIT) Actually, pod-to-nodeport with `externalTrafficPolicy: Local` [doesn't currently work in most cases anyway](https://github.com/kubernetes/kubernetes/pull/106497/commits/1da8903a09d9a12c660397a33a0e98fe973d121e#diff-92f41de420a097c0419a3cb53cf176a459a742c39f3fa482ab48b1c2fa6fd084R5463-R5476).
    - we could potentially make traffic to NodePorts on 127.0.0.1 be "internal" and traffic to NodePorts on other IPs be "external"...
- Pod/Node to ExternalIP or LoadBalancer IP would always work, even if the service was `internalTrafficPolicy: Local, externalTrafficPolicy: Local` and had no local endpoints. (The mental model here is that for LB IPs, Local traffic policy is purely an internal implementation detail, and clients never have to think about it. As opposed to for NodePorts, where Local traffic policy is an explicitly visible aspect of their behavior, which clients _do_ need to think about.)
    - What exact set of endpoints gets used in the "eTP Local but no local endpoints" case? Is it guaranteed to be the same set of possible endpoints that might be used if an external client had connected to the LB IP? eg, do we ignore topology and possibly use terminating endpoints? Or do we use "the set of endpoints that we would have used if you connected to the service in a way that used Cluster traffic policy"?
- All other connection paths should be unchanged.

### PR DESCRIPTION
**Pull request by danwinship**
#### What type of PR is this?
/kind documentation

#### What this PR does / why we need it:
Updates API docs based on discussion in #108526 and changes in #106497.

#### Which issue(s) this PR fixes:
Fixes #108526

#### Does this PR introduce a user-facing change?
(no, but it documents changes that were already release-noted in #106497)
```release-note
NONE
```

/sig network
/triage accepted
/priority important-soon
/milestone v1.24
/assign @thockin

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -3853,12 +3859,16 @@ type ServiceSpec struct {
 	// +optional
 	LoadBalancerSourceRanges []string
 
-	// externalTrafficPolicy denotes if this Service desires to route external
-	// traffic to node-local or cluster-wide endpoints. "Local" preserves the
-	// client source IP and avoids a second hop for LoadBalancer and Nodeport
-	// type services, but risks potentially imbalanced traffic spreading.
-	// "Cluster" obscures the client source IP and may cause a second hop to
-	// another node, but should have good overall load-spreading.
+	// externalTrafficPolicy describes the endpoint-selection policy for traffic
```
Comment (thockin) [inline pkg/apis/core/types.go:3862]: It's not really about "received from outside the cluster" but "received at external-facing entrypoints" ?

File: pkg/apis/core/types.go [hunk 2]
```diff
@@ -3853,12 +3859,16 @@ type ServiceSpec struct {
...
+	// received from outside the cluster on one of the Service's external IPs/ports
+	// (NodePorts, ExternalIPs, and LoadBalancer IPs). "Cluster" (the default) uses
+	// the standard behavior of routing to all endpoints (possibly modified by
+	// topology and other features), which should have good overall load-spreading but
+	// normally obscures the client source IP. "Local" delivers traffic only to
+	// node-local endpoints, avoiding a second hop to another node for traffic that
+	// was already load-balanced externally, and preserving the client source IP. Note
+	// that externalTrafficPolicy may not apply to traffic sent to an external IP/port
+	// from a client inside the cluster.
```
Comment (thockin) [inline pkg/apis/core/types.go:3871]: Should we detail the sepcial cases here?

**Review by thockin — COMMENTED** [review]
Great edit overall

File: pkg/apis/core/types.go (hunk 1 above)
Comment (danwinship) [reply pkg/apis/core/types.go:3862]: but with the special case for pod and node traffic, it ends up being that externalTrafficPolicy only affects traffic which both comes from outside the cluster and ends up an external-facing entrypoing.

(The internal case is not currently parallel; see https://github.com/kubernetes/kubernetes/issues/108526#issuecomment-1086080420)

File: pkg/apis/core/types.go (hunk 2 above)
Comment (danwinship) [reply pkg/apis/core/types.go:3871]: Is the API documentation really the right place to dive into the semantics of "pod to same-node NodePort" vs "pod to remote NodePort", or AWS-style load balancers vs GCP-style load balancers?

(Is the difference in behavior between pod-to-same-node-NodePort and pod-to-remote-NodePort even canonically part of the definition of Local traffic policy? Or is that just the behavior of iptables kube-proxy, where we care about preserving the specific backward-compatible behavior _in iptables kube-proxy_ but don't necessarily care if other impementations behave precisely the same way? Considering, for example, that in some plugins with a more "centrally-programmed" data plane (like ovn-kubernetes), it's not only _possible_, but actually _easier_ to make pod-to-NodePort always behave the same way regardless of whether it's same-node or different-node.)

I guess the API docs should describe the behavior that is intentional and reliable and encouraged... so I'd say that we don't say anything about pod/node to NodePort, and that for pod/node to externalIP/LBIP it always has Cluster semantics.

File: pkg/apis/core/types.go (hunk 1 above)
Comment (thockin) [reply pkg/apis/core/types.go:3862]: if a pod hits a node-port on a different node, it is subject to eTP (right?) but isn't "received from outside the cluster".

Words are hard.  I personally find it easier to think of it as "external-facing inputs" and "a couple special cases".  I am open to anything that makes this clearer. :)

File: pkg/apis/core/types.go (hunk 2 above)
Comment (thockin) [reply pkg/apis/core/types.go:3871]: I can buy the "describe the behavior that is intentional" theme.  We already have implementation-defined things like localhost and we don't call that out, either.  Pod->nodeport is a weird behavior anyway. Your approach SGTM

File: pkg/apis/core/types.go (hunk 1 above)
Comment (aojea) [reply pkg/apis/core/types.go:3862]: ... seems the node is the boundary?

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve

**Pull request merged**